policy/daemon.go

```diff
@@ -168,93 +172,69 @@ func (p *policyd) Update(ctx context.Context) error {
 		return err
 	}
 
-	rp.StartExpired(ctx, p.policyExpiredDuration).
-		EnableExpiredHook().
-		SetExpiredHook(func(ctx context.Context, key string) {
-			//key = <domain>:role.<role>
-			p.fetchAndCachePolicy(ctx, p.rolePolicies, strings.Split(key, ":role.")[0])
-		})
-
-	p.rolePolicies, rp = rp, p.rolePolicies
-	rp.Stop()
-	rp.Clear()
-
+	glg.Info("Policy updated")
 	return nil
 }
 
 // CheckPolicy checks the specified request has privilege to access the resources or not.
 // If return is nil then the request is allowed, otherwise the request is rejected.
 func (p *policyd) CheckPolicy(ctx context.Context, domain string, roles []string, action, resource string) error {
-	ech := make(chan error, 1)
-	cctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	ch := make(chan error, 1)
 
 	go func() {
-		defer close(ech)
 		wg := new(sync.WaitGroup)
 		for _, role := range roles {
-			dr := fmt.Sprintf("%s:role.%s", domain, role)
 			wg.Add(1)
-			go func(ch chan<- error) {
+			go func(r string) {
 				defer wg.Done()
-				select {
-				case <-cctx.Done():
-					ch <- cctx.Err()
-					return
-				default:
-					asss, ok := p.rolePolicies.Get(dr)
-					if !ok {
-						return
-					}
-
-					for _, ass := range asss.([]*Assertion) {
-						glg.Debugf("Checking policy domain: %s, role: %v, action: %s, resource: %s, assertion: %v", domain, roles, action, resource, ass)
-						select {
-						case <-cctx.Done():
-							ch <- cctx.Err()
-							return
-						default:
+				if roleAsss, ok := p.rolePolicies.Get(domain); ok {
+					if asss, ok := roleAsss.(map[string][]*Assertion)[r]; ok {
+						for _, ass := range asss {
+							glg.Debugf("Checking policy domain: %s, role: %v, action: %s, resource: %s, assertion: %v", domain, roles, action, resource, ass)
 							if strings.EqualFold(ass.ResourceDomain, domain) && ass.Reg.MatchString(strings.ToLower(action+"-"+resource)) {
 								ch <- ass.Effect
```

Comment:
I think we will have deadlock here, when multiple deny policy match.
1. ech size = 1
2. read from ech only 1 times
3. when, write more than 2, the go func cannot return, wg.Wait() will block forever

better to keep the old logic first

Reply:
refer to: https://github.com/yahoojapan/athenz-authorizer/issues/30